netket/machine/torch.py

```diff
@@ -76,50 +85,70 @@ def n_par(self):
     def log_val(self, x, out=None):
         if len(x.shape) == 1:
             x = x[_np.newaxis, :]
-
         batch_shape = x.shape[:-1]
 
         with _torch.no_grad():
             t_out = self._module(_torch.from_numpy(x)).numpy().view(_np.complex128)
-
         if out is None:
             return t_out.reshape(batch_shape)
-
         _np.copyto(out, t_out.reshape(-1))
 
         return out
 
     def der_log(self, x, out=None):
+
         if len(x.shape) == 1:
             x = x[_np.newaxis, :]
         batch_shape = x.shape[:-1]
         x = x.reshape(-1, x.shape[-1])
 
+        if out is None:
+            out = _np.empty([x.shape[0], self._n_par], dtype=_np.complex128)
         x = _torch.tensor(x, dtype=_torch.float64)
-        out = x.new_empty([x.size(0), 2, self._n_par], dtype=self._parameters[0].dtype)
-        m = self._module(x)
 
-        for i in range(x.size(0)):
-            dws_real = _torch.autograd.grad(
-                m[i, 0], self._parameters, retain_graph=True
-            )
-            dws_imag = _torch.autograd.grad(
-                m[i, 1], self._parameters, retain_graph=True
-            )
-            _torch.cat([dw.flatten() for dw in dws_real], out=out[i, 0, ...])
-            _torch.cat([dw.flatten() for dw in dws_imag], out=out[i, 1, ...])
-
-        out_complex = _np.zeros((out.size(0), out.size(2)), dtype=_np.complex128)
-        out_complex = out[:, 0, :].numpy() + 1.0j * out[:, 1, :].numpy()
-
-        return out_complex.reshape(
-            tuple(list(batch_shape) + list(out_complex.shape[-1:]))
-        )
+        if self._use_backpack:
+
+            def write_to(dst):
+                dst = _torch.from_numpy(dst)
+                i = 0
+                for gb in (
+                    p.grad_batch.flatten(start_dim=1)
+                    for p in self._module.parameters()
+                    if p.requires_grad
+                ):
+                    dst[:, i : i + gb.shape[1]].copy_(gb)
+                    i += gb.shape[1]
+
+            m_sum_real = self._module(x)[:, 0].sum(dim=0)
```

Comment:
It'd be nice to check what is more efficient: to do the reduction to scalar here or to provide `grad` argument to `backward`. It doesn't have to happen now though, just create an issue if you don't have time to check it yourself.

Reply:
Inside pytorch when a grad is provided to backward the input is also reduced to a scalar first, so I think it's just a problem of different coding styles. I prefer to write the sum explicitly to make it easier to read.